pkg/configurator/validating_webhook.go

```diff
@@ -0,0 +1,363 @@
+	var configMap corev1.ConfigMap
+	if _, _, err := deserializer.Decode(req.Object.Raw, nil, &configMap); err != nil {
+		log.Error().Err(err).Msgf("Error unmarshaling request to configmap in namespace %s", req.Namespace)
+		return webhook.AdmissionError(err)
+	}
+
+	log.Trace().Msgf("Validation request: (new object: %v) (old object: %v)", string(req.Object.Raw), string(req.OldObject.Raw))
+
+	// Default response
+	resp := &v1beta1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{Reason: ""},
+		UID:     req.UID,
+	}
+
+	// Allow if not a change to osm-config
+	if req.Name != constants.OSMConfigMap {
+		return resp
+	}
+
+	checkDefaultFields(configMap, resp)
+	whc.validateFields(configMap, resp)
+
+	return resp
+}
+
+// checkDefaultFields checks that all default fields for osm-config exist
+func checkDefaultFields(configMap corev1.ConfigMap, resp *v1beta1.AdmissionResponse) *v1beta1.AdmissionResponse {
+	data := make(map[string]struct{})
+	for field := range configMap.Data {
+		data[field] = struct{}{}
+	}
+
+	for def := range defaultFields {
+		if _, ok := data[def]; !ok {
+			reasonForDenial(resp, doesNotContainDef, def)
+		}
+	}
+	return resp
+}
+
+// validateFields checks whether the configmap field values are valid and rejects as necessary
+func (whc *webhookConfig) validateFields(configMap corev1.ConfigMap, resp *v1beta1.AdmissionResponse) *v1beta1.AdmissionResponse {
+	for field, value := range configMap.Data {
+		if !checkBoolFields(field, value, boolFields) {
+			reasonForDenial(resp, mustBeBool, field)
+		}
+		if field == "envoy_log_level" && !checkEnvoyLogLevels(field, value) {
+			reasonForDenial(resp, mustBeValidLogLvl, field)
+		}
+		if field == "service_cert_validity_duration" {
+			_, err := time.ParseDuration(value)
+			if err != nil {
+				reasonForDenial(resp, mustBeValidTime, field)
+			}
+		}
+		if field == "tracing_address" && !matchAddrSyntax(field, value) {
+			reasonForDenial(resp, mustFollowSyntax, field)
+		}
+		if field == "tracing_port" {
+			portNum, err := strconv.Atoi(value)
+			if err != nil {
+				reasonForDenial(resp, mustbeInt, field)
+			}
+			if portNum < 0 || portNum > maxPortNum {
+				reasonForDenial(resp, mustBeInPortRange, field)
+			}
+		}
+	}
+
+	defConfigMap, _ := whc.kubeClient.CoreV1().ConfigMaps(whc.osmNamespace).Get(context.TODO(), constants.OSMConfigMap, metav1.GetOptions{})
+
+	for metadataAnnotation, val := range configMap.ObjectMeta.Annotations {
+		if defConfigMap.Annotations[metadataAnnotation] != val {
+			reasonForDenial(resp, cannotChangeMetadata, metadataAnnotation)
+		}
+	}
+	for metadataLabels, val := range configMap.ObjectMeta.Labels {
+		if defConfigMap.Labels[metadataLabels] != val {
+			reasonForDenial(resp, cannotChangeMetadata, metadataLabels)
+		}
+	}
+	return resp
+}
+
+// matchAddrSyntax checks that string follows hostname syntax
+// example hostname address: abc123.1cd23.efg
+func matchAddrSyntax(configMapField, configMapValue string) bool {
```

Comment:
Could you add a table driven unit this for this function with different inputs to see clearly what is allowed and what's not.

Also why is this check so necessary to have?

Reply:
it's not necessary, i'll go ahead and remove it